Allow dynamic service registration and ability for local processes to register service names

grondo opened the pull request:
This PR adds basic support for dynamic service registration via `service.add` and `service.remove` RPCs as suggested in #1689. Support for `service.add/remove` is also added to the local connector which allows locally connected clients to register services with the broker.

The more ambitious work to allow agents to register service names that shadow existing services was parked for now in the interest of completing required functionality.

Still need to add tests, but basic support is here for comments.

grondo commented:
example:
```lua
#!/usr/bin/lua
local flux = require 'flux'
local f = assert (flux.new())
local inspect = require 'inspect'

local function printf (...)
    io.stdout:write (string.format (...))
end

local mh, err = f:msghandler {
    pattern = "foo.ping",
    msgtypes = { flux.MSGTYPE_REQUEST },
    handler = function (f, msg, mh) msg:respond { seq = msg.data.seq } end
}
if not mh then error (err) end

local mh, err = f:msghandler {
    pattern = "foo.*",
    msgtypes = { flux.MSGTYPE_RESPONSE },
    handler = function (f, msg, mh)
        local resp = msg.data
        if resp.errnum then
            error (msg.tag .. ": "..resp.errnum)
        end
        local seq = resp.seq
        printf ("%s seq=%d\n", msg.tag, seq)
        if seq == 10 then
            f:reactor_stop()
            return
        end
        f:send ("foo.ping", { seq = seq + 1 })
    end
}
if not mh then error (err) end

-- Register with local connector for the "foo" service:
local rc, err = f:rpc ("service.add", { service = "foo" })
if not rc then error (err) end
printf ("Added service %s\n", "foo")

-- Send the first ping and enter reactor
assert (f:send ("foo.ping", { seq = 1 }))
f:reactor()

-- Deregister service.foo (we could also let disconnect deregister us)
local rc, err = f:rpc ("service.remove", {service = "foo"})
if not rc then error (err) end
```

```shell
$ src/cmd/flux start lua t.lua 
Added service foo
foo.ping seq=1
foo.ping seq=2
foo.ping seq=3
foo.ping seq=4
foo.ping seq=5
foo.ping seq=6
foo.ping seq=7
foo.ping seq=8
foo.ping seq=9
foo.ping seq=10
```

garlick commented:
Nice!

Couple of quick comments:
* we should drop the `service_name` thing from RFC 6 like it never happened!  :-)
* could the `service_` message handlers live in service.c?   e.g. with a `service_switch_set_flux()` function called from the broker to register them once the local `flux_t` handle is available?

grondo commented:
> we should drop the service_name thing from RFC 6 like it never happened! :-)

Good point! I didn't think to check the RFC.

> could the service_ message handlers live in service.c? e.g. with a service_switch_set_flux() function called from the broker to register them once the local flux_t handle is available?

This would have been my preference and was my initial approach. However, I ran into awkward trouble quickly since the module code and "service switch" code are in different abstractions. i.e. it seemed like trying to put the `service.*` handling code, which needs to access the module hash, was breaking these abstractions that were already set up. There is a comment to that effect on the msg handlers for those RPCs:

```C
/* Dynamic service registration.
 * These handlers need to appear in broker.c so that they have
 *  access to broker internals like modhash
 */
static void service_add_cb (flux_t *h, flux_msg_handler_t *w,
                            const flux_msg_t *msg, void *arg)
{
```

At some future point, we will probably want to have the "modservice" and "service.*" code better integrated. Even now it might be nice to be able to include the service names for loaded modules, but the extension module RFC nor the existing code easily allows that. Fixing that seemed like more than we would want to tackle for this PR though.

garlick commented:
> I ran into awkward trouble quickly since the module code and "service switch" code are in different abstractions.

Uh, sorry, I should have looked more closely before speaking.  That cleanup's probably for another day and your approach makes sense!

grondo commented:
> I think probably a couple of API functions ought to be added like:

I thought about that, but I kind of lean toward not adding a new API function for every RPC. We already have a nice rpc interface, and the API is growing very rapidly, which then requires language bindings. I figured at this point it was unnecessary to wrap the 3 lines of code to send add/remove RPCs -- an API interface could be added later when we figure out if this direction is useful.

> It seems like client->services should be a zlist? Or, since that's for disconnect handling, would it maybe be possible to eliminate it entirely and just do a linear search through the ctx->services looking for a match for the client_t pointer? (presuming there will be relatively few services registered this way)

It is much easier to maintain the set of service names in use by a client as a set, not a list. What is the benefit of the extra code to do linear searches through zlists exactly? Sorry if I'm missing some obvious drawback of a hash.

grondo commented:
Actually you make a good point about just dropping `client->services` hash and doing a linear search at disconnect. This is what the broker already does anyway, and it will remove a member from each client which would largely be unused.

garlick commented:
> I kind of lean toward not adding a new API function for every RPC

Were you thinking this is for internal use only at this point?  

> It is much easier to maintain the set of service names in use by a client as a set, not a list. What is the benefit of the extra code to do linear searches through zlists exactly? Sorry if I'm missing some obvious drawback of a hash.

Oh sorry, maybe I misread, it looked like the hash was being used at a list, and the hash pre-allocates 128 slots (or something like that, so it's a bit heavy if it's not ever being used as a hash. I'll have another look at that.

grondo commented:
Also added here is an experimental update to pymod to dynamically register a service name based on the name of the script being loaded, instead of just `pymod`. (There is also commented code to *remove* the `pymod` service name, which could theoretically let multiple `pymod` scripts be loaded at the same time. However, this isn't safe in practice, nor is there a way to remove multiple modules with the same name presently)

The `echo.py` example is also updated to take advantage of the new pymod feature, and it actually works:

```
grondo@peep:~/git/f.git$ flux module load pymod --verbose --path=$(pwd)/src/modules/pymod echo
grondo@peep:~/git/f.git$ flux dmesg | grep pymod
2018-10-24T03:27:05.753254Z broker.debug[0]: insmod pymod
2018-10-24T03:27:05.770080Z pymod.info[0]: loading python module named: echo
grondo@peep:~/git/f.git$ cat echo.lua 
#!/usr/bin/lua
local f = assert (require 'flux'.new())
local resp = assert (f:rpc ("echo.foo", { data = "foo" }))
print (require 'inspect' (resp))

grondo@peep:~/git/f.git$ lua echo.lua 
{
  data = "foo"
}
```

garlick commented:
The pymod change is pretty cool! 

> I hadn't considered it, but in the timeframe of the current sprint, yes possibly. flux_rpc is exported via the API though, correct? It seems like we could take advantage of that sometimes instead of wrapping every simple RPC in a new API call? I'm probably missing some perspective at this point though..

If you feel strongly about it, I'm fine with that for now.  My general feeling is, for a new service that _might_ be externally useful, it seems like a good compromise position is to provide the API call with header-only documentation.  It makes the thing easier to figure out for users, and allows the expected payloads, etc. to change without the need to go find users and fix them.

I agree API bloat is a problem though.  The zeromq people have a way of tagging interfaces as experimental, which allows them to get rid of things that don't work out over a few release cycles.  Maybe something like that's a way to be more deliberate about API growth?

grondo commented:
> If you feel strongly about it, I'm fine with that for now. My general feeling is, for a new service that might be externally useful, it seems like a good compromise position is to provide the API call with header-only documentation. It makes the thing easier to figure out for users, and allows the expected payloads, etc. to change without the need to go find users and fix them.

Well I don't feel too strongly about it, but each time I wrap a simple RPC in an API function, I kind of wonder what the point of the nice `flux_rpc_pack` interface is, since no external users will have occasion to use it. Sometimes if feels like it would be better to have a nice way to document protocols and allow us the speed and flexibility of rapidly developing RPCs, rather than adding that extra level of C api that then needs to be separately maintained as the protocol evolves. With the C api you also require bindings to be maintained instead of using an existing, possibly nicer RPC interface.

However, I can see the benefit of abstracting the protocol away from (most) users, and I don't feel too strongly so I can add functions to the API (was mainly just defending my position *not* adding them to the initial PR)

grondo commented:
> Also added here is an experimental update to pymod

I forgot to mention that there is a problem with `pymod` on systems where `_cffi_backend.so` is not linked against `libpython*.so`. In this case, `pymod` will not be able to load and run a python script because libpython symbols won't have visibility in the script, so `import *` will fail with an undefined symbol error.

To workaround, you have to dlopen the correct `libpython*.so` so you can promote its flags to `RTLD_GLOBAL`. (This is the same issue we hit with librdl loading Lua modules when it was loaded in the context of `sched.so`)

On TOSS, at least the default `_cffi_backend.so` is linked correctly with libpython2.7.so. However, on Ubuntu 18.04 it doesn't appear to have proper linking so this workaround would be required. The tricky part, though, is finding the right name to dlopen()... Not sure there is a good way to do that.

grondo commented:
Latest push enables pymod on distros where _cffi_backend.so isn't linked with libpython by dlopening the correct libpython (as determined during configure) with `RTLD_GLOBAL`.

This allows a sanity test for pymod to be added to sharness testsuite.

The last bit here is to add a test for connector-local dynamic services.

grondo commented:
I rebased this on top of #1772 since that PR makes it easier to add the `pymod` test unconditionally.

I've also added a new test under `python/t1000-service-add-remove.py` which test the `service.add` `service.remove` interface. I kind of stumbled along writing the test, so would appreciate a gander by @SteVwonder or @trws.

SteVwonder commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,91 @@
+            self.assertEqual (msg, p)
+            f.reactor_stop (f.get_reactor())
+
+        w2 = self.f.msg_watcher_create (cb, FLUX_MSGTYPE_RESPONSE,
```
I believe this watcher needs to be started before the reactor is entered.

SteVwonder commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,91 @@
+    def test_005_service_rpc_enosys (self):
+        fut = self.f.flux_rpc ("foo.echo", "{}", FLUX_NODEID_ANY, 0)
+	with self.assertRaises(Exception) as e:
+            self.f.flux_future_get (fut, ffi.NULL)
```
I think here you can take advantage of the default arguments in the rpc bindings and it still work the same.
```python
        fut = self.f.rpc_create ("foo.echo")
        with self.assertRaises(EnvironmentError) as e:
            fut.get ()
        self.assertEqual (e.exception.errno, errno.ENOSYS)
```

grondo replied:
is there a `.get()` method on futures in the python bindings yet? `fut.get()` doesn't seem to work for me:
```
      AttributeError: cdata 'struct flux_future *' points to an opaque type: cannot read fields
```

trws replied:
We haven't re-worked the futures yet, there isn't a generic future at all at this point unfortunately.

trws replied:
@SteVwonder is right, but I think I see the issue.  If you call `f.rpc_create()` it returns a python wrapper that adds the `get()` method.  The raw `flux_rpc ()` call returns something like a `cffi.ctype("flux_future_t *")` which doesn't have the wrapper class to provide methods.

grondo replied:
Got it, thanks!

SteVwonder commented on t/t9001-pymod.t:
```diff
@@ -0,0 +1,40 @@
+'
+
+# Test can be a tad "racey" at times if grace period is too low, up to
+# 5 seconds to ensure joblog can be generated properly.
```
Accidental copy-paste carryover?

grondo commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,91 @@
+  return 2
+
+def service_add (f, name):
+    future = f.flux_service_register (name)
```
Ah, ok so `flux_` prefix can be dropped from the automatic ffi functions?

grondo commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,91 @@
+        self.f.close()
+
+    def test_001_register_service (self):
+	rc = service_add (self.f, "foo")
```
Thanks! I will remove all tabs from the file.

grondo commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,91 @@
+        self.assertEqual(rc, 0, msg="flux_service_register ('foo')")
+
+    def test_002_service_add_eexist (self):
+	with self.assertRaises(Exception) as e:
```
ok, great point!

trws commented:
It’s in the prefix list that gets auto-added by the wrappers for handles, so yeah.

On 26 Oct 2018, at 14:14, Mark Grondona wrote:

@grondo commented on this pull request.

________________________________

In t/python/t1000-service-add-remove.py<https://github.com/flux-framework/flux-core/pull/1753#discussion_r228667485>:

> +import unittest
+import errno
+import flux.core as core
+from flux.message import Message
+from flux.core.inner import ffi
+from flux.constants import ( FLUX_MSGTYPE_REQUEST,
+                             FLUX_MSGTYPE_RESPONSE,
+                             FLUX_NODEID_ANY )
+
+from subflux import rerun_under_flux
+
+def __flux_size():
+  return 2
+
+def service_add (f, name):
+    future = f.flux_service_register (name)


Ah, ok so flux_ prefix can be dropped from the automatic ffi functions?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/flux-framework/flux-core/pull/1753#discussion_r228667485>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAoStVGTzsdTK0MOenshn2MFnIwaoBm_ks5uo3s2gaJpZM4X2Mz5>.

grondo commented:
Rebased, addressed @SteVwonder's comments I believe, and got the `python/t1000-service-add-remove.py` test working. *However*, this required the workaround discussed in #1776 (`connector-local` module always enables routes on requests), and I'm not sure that is acceptable. (However, I wanted to get the test working in python, and not have to fall back to Lua or C to test)

grondo commented:
Hm, the distcheck builders are getting a failure in the pymod test:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named flux
not ok 2 - pymod echo.py function works
```

This might mean that PYTHONPATH is not quite correct in the VPATH builds (however, all the other python tests work, so it might be more subtle than that. Probably just affects the environment of the broker itself?)

SteVwonder commented:
Yeah, that is perplexing considering that [pymod successfully imports `flux.core`](https://github.com/grondo/flux-core/blob/dynamic-services/src/modules/pymod/py_mod.c#L179) in order to get `mod_main_trampoline` and import/run `echo.py`.

grondo commented:
I just tried an experiment and changed the PYTHONPATH exported in `t/Makefile.am` from using `$(top_srcdir)` and `$(top_builddir)` to `$(abs_top_srcdir)` and `$(abs_top_builddir)` and that seems to work around the issue.

Something during import seems to be removing the relative paths (`sysconfig` module?)

grondo commented:
After #1766 goes in I'll update the invocations of `python` in this PR to `flux python`

grondo commented:
Rebased, found and fixed an issue with MessageWatcher.destroy method, and updated the test to include a test of connector-local disconnect handling.

garlick commented:
Great, is there anything remaining that needs to happen before this can go in?

grondo commented:
I'm still looking at low coverage (might be an unused function)

Also, could you approve the workaround in fa1a6e4e18180b26592a74d61920143b35846e2e?

garlick commented:
Just commented in there - and it shows as a little dialog bubble rather than inline.  Hmm.  Anyway, LGTM.

grondo commented:
> Just commented in there - and it shows as a little dialog bubble rather than inline. Hmm. Anyway, LGTM.

Ok, thanks! Let me just test & push a change to remove an apparently unused function then this should be ready.

grondo commented on t/python/t1000-service-add-remove.py:
```diff
@@ -0,0 +1,131 @@
+        self.assertEqual(rc, 0)
+
+    def test_002_service_add_eexist(self):
+        with self.assertRaises(Exception) as e:
```
No problem, not sure how I missed that!

grondo commented on src/modules/pymod/py_mod.c:
```diff
@@ -97,6 +98,47 @@ static struct optparse_option opts[] = {
+    /* First deregister the "pymod" name. This should allow another
+     *  script to be loaded after we return
+     */
+#if 0
```
I would like to leave it there because it is instructive for the future. If you feel strongly that it should be removed, however, I can remove it.

SteVwonder replied:
Sounds good to me :+1:

grondo commented:
Thanks for the review @SteVwonder! Addressed your comments, squashed and force-pushed.

grondo commented:
I actually ended up removing that code that was `#ifdef`'d out. It may have been instructive, but probably a bad example anyway

grondo commented:
Thanks!
